Add WebAssembly support and the ability to Dial from browsers

albrow opened the pull request:
This PR is still a work in progress as there are some details to work out (especially around error handling). I should probably also do one more round of cleanup. However, the basics of the transport are working. This PR also includes an integration test which covers basic communication over a stream between a server (running the vanilla Go transport) and a headless Chrome browser (running the WebAssembly/JavaScript version).

After doing some more investigation, I think it makes more sense to introduce direct WebAssembly support here rather than adding it to the `gorilla/websocket` dependency. The main reason is that the `gorilla/websocket` API has a relatively large surface area, their own set of configuration options, and their own way of establishing connections that don't translate well to Wasm/JavaScript. The subset of the `gorilla/websocket` API that is used in this package is much easier to translate.

The `Listen` method is not supported Wasm/JavaScript environments and calling it will always return an error. Browsers cannot listen for incoming WebSocket connections; they can only dial.

I'm open to suggestions about how to organize the files, but right now I'm using the file name suffixes `_go` and `_js` to denote files that only work in vanilla Go or JavaScript/Wasm environments, respectively. Files without any suffix work in both environments.

So for example, __conn.go__ has been split into three files:
  - __conn.go__ contains code that is shared between vanilla Go implementation and the JavaScript bindings.
  - __conn_go.go__ contains the vanilla Go implementation.
  - __conn_js.go__ contains the JavaScript bindings.

While the implementations defer, the exported API in `_go` and `_js` files is the same and a lot of code is shared.

albrow commented:
CI is failing because it uses Go 1.11. My PR is targeting the latest go version as of now which is 1.12. Can we update Travis to use the latest version? If you want to run the browser integration tests in Travis, we might also need to install Chrome.

albrow commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+	closeSignal    chan struct{}
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
```
I would appreciate it if someone could take a closer look at this implementation of `Read`. I'm not super well-versed in the semantics/behavior libp2p expects.

albrow commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+
+func (c *Conn) LocalAddr() net.Addr {
+	// TODO(albrow): is there some way to get the local address?
+	return NewAddr("0.0.0.0:0")
```
AFAIK the JavaScript WebSocket API does not expose any way to get the local address/port. Is that going to be a problem? I'm using the workaround of just returning `0.0.0.0:0` but maybe there is a better way to indicate that the local address is unknown?

Stebalien replied:
This should be fine.

Stebalien commented:
Is there any reason to merge this _here_ instead of to simply create an entirely new transport?  This is basically just a different implementation of the websocket transport and there isn't any shared code.

We'd then conditionally import the correct transport from go-libp2p.

albrow commented:
@Stebalien there is a decent amount of shared code. Every file without a build tag at the top is shared. This includes:

- __addrs.go__
- __addrs_test.go__
- __websocket.go__
- __conn.go__

In terms of functionality, the shared code:

- Specifies the multiaddress format for the transport and converts to/from a WebSocket `net.Addr`.
- Specifies the protocol and codec (`WsProtocol` and `WsCodec`).
- Uses a `tptu.Upgrader` to upgrade a multiaddr-net connection into a full libp2p-transport connection.
- Implements most of the methods in the `Transport` interface: `CanDial`, `Protocols`, `Proxy`, and `Dial` (partially at least; the environment-specific `maDial` method does most of the heavy-lifting).

There is a benefit to keeping this shared code in one repository so that if we ever need to change it in the future, we only need to change it in one place.

There is another big advantage to putting the WebAssembly-related code here: it allows us to write integration tests that ensure compatibility. The goal here is to make the API and behavior the same regardless of whether you are compiling to WebAssembly or a binary executable, and integration tests are an important part of making sure we adhere to that goal. It's possible to have integration tests if the code lives in two different repositories, but again that would make it more difficult to manage if we ever made changes in the future.

Finally, keeping this code in the same place makes it easier for developers who want to enable/disable specific transports instead of just using the libp2p defaults. In that case, they can just import `github.com/libp2p/go-ws-transport` and it will work in both vanilla Go and JavaScript/Wasm environments. This the same approach that we followed in [go-libp2p-webrtc-direct](https://github.com/libp2p/go-libp2p-webrtc-direct), [github.com/pion/webrtc](https://github.com/pion/webrtc) and it works really well. If the code lived in two different repositories, developers who want to build isomorphic apps would need to create two different files with different build tags and import two different packages.

Stebalien commented:
My general concern is that it means anyone wanting to modify the normal go transport needs to make sure they don't accidentally break the wasm transport. However, you're right. It's a lot simpler to explain "here's the websocket transport" instead of "here's the websocket transport for platform A, here's the websocket transport for platform B".

Note, in practice:

1. The same config won't quite work as most users will want to enable TCP (which can't work in WASM).
2. We can select the correct defaults for the platform in go-libp2p.

However, this is still probably the best way to do it.

Stebalien commented:
> I'm open to suggestions about how to organize the files, but right now I'm using the file name suffixes _go and _js to denote files that only work in vanilla Go or JavaScript/Wasm environments, respectively.

Maybe `_browser` versus `_system` (or `_native`)? This doesn't technically involve any javascript.

Stebalien commented:
> The Listen method not supported Wasm/JavaScript environments and calling it will always return an error. Browsers cannot listen for incoming WebSocket connections; they can only dial.

That should be fine. The user will just get an error on construction if they ask the browser to listen on a websocket address.

albrow commented:
> My general concern is that it means anyone wanting to modify the normal go transport needs to make sure they don't accidentally break the wasm transport.

I understand that concern and I'm happy to help field any issues that might come up 😄This package is critical for my project at work so I can commit some time to help maintain it. I'd also be happy to add some documentation here to help get contributors get oriented since most people will not have prior experience with Wasm.

> The same config won't quite work as most users will want to enable TCP (which can't work in WASM).

Yeah that's a good point. There will be *some* difference in configuration but it's still easier if you don't have to import different packages for different platforms. And you're right it's probably still a good idea to have go-libp2p choose the appropriate defaults.

albrow commented:
I should've shared this before. For anyone unfamiliar with how WebAssembly works in Go, these are good resources to get started:

- [The Go WebAssembly Wiki](https://github.com/golang/go/wiki/WebAssembly)
- [The syscall/js package](https://golang.org/pkg/syscall/js/)

albrow commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	// c.mut.Lock()
```
Another reason I called this PR "WIP" is because I need to go through and make sure we are acquiring and releasing a lock when needed. Will take a little bit of careful thought to avoid both race conditions and deadlocks.

albrow replied:
I addressed this by adding two separate locks. One controls the callbacks for event handlers. We need to call `Release` on these whenever the connection is closed. The other is specifically for controlling access to the buffer which represents incoming data. Let me know what you think.

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, io.EOF
```
This should only return `io.EOF` if the connection was successfully closed. Is there any way to distinguish between some error condition and "closed"?

albrow replied:
`checkOpen` only returns an error if the connection is closed. I still need to finish handling errors in some cases (e.g. if the ["onerror" event](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/onerror) is triggered) but I will make sure we return any unexpected errors.

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	// Not supported
+	return nil
```
These should return `os.ErrNoDeadline`.

albrow replied:
@Stebalien I tried making this change but it resulted in an error when trying to actually use the transport in a browser:

```
failed to negotiate security stream multiplexer: file type does not support deadline
```

I believe the error is coming from https://github.com/libp2p/go-libp2p-mplex or https://github.com/libp2p/go-libp2p-transport-upgrader.

Stebalien replied:
Ah. Can you leave a todo for now? We should both add support for this _and_ fix the multiplexer interfaces to take a context (we currently use deadlines to avoid this issue).

albrow replied:
👍 I added a TODO.

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+	loadEndFunc := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			text := args[0].Get("srcElement").Get("result").String()
+			textChan <- text
```
`textChan` should just have a buffer of one instead of launching a goroutine.

albrow replied:
One quirk of the `syscall/js` package is that you can't block inside of any JavaScript callback (i.e. anything constructed with `js.FuncOf`). Blocking inside a JavaScript callback blocks the entire JavaScript event loop and no other callbacks can be processed.

The latest version of this PR uses a buffer of 100 and a separate goroutine whose sole responsibility is reading from the channel. It is still technically possible to block inside the JavaScript callback but that would only happen if the buffered channel were full, which is incredibly unlikely. Let me know what you think of this approach.

Stebalien replied:
Here, a buffer of one would have been fine.

> Let me know what you think of this approach.

I think we should use a lock, buffer, and a tight critical section where we write into that buffer.

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+	defer loadEndFunc.Release()
+	reader.Call("addEventListener", "loadend", loadEndFunc)
+	reader.Call("readAsText", blob)
+	return <-textChan
```
Error handling? What if "loadend" never fires?

albrow replied:
If JavaScript throws, we can catch it with `defer` and `recover`. I can add that here like I did in a few other functions.

However, I'm not sure how to handle the case of "loadend" not firing. We can use a `select` and `time.After` to timeout after some amount of time if "loadend" doesn't fire, but what should we do next? Trying to keep the connection open could mean missing data or out-of-order reads. Maybe we should just close the connection if this happens?

Stebalien replied:
If loadend fires on error, we should be fine. Otherwise, we'll need to register an event listener for that error and select on it and loadend.

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
```
This could deliver messages out of order. We need some kind of in-order buffer. We can do that by either:

1. Spinning up a dedicated goroutine to read from this channel.
2. Ideally, just writing directly into a shared buffer (protected by a mutex). Availability of data to read could be signaled by writing an empty `struct{}{}` to a flag channel.

albrow replied:
@Stebalien take a look at the latest version and let me know what you think. Just need to figure out how and when to reset the buffer.

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+	if c.currData == nil {
+		select {
+		case data := <-c.incomingData:
+			c.currData = bytes.NewBuffer(data)
```
Actually, ideally this would re-use the same reader (and just reset it as necessary) instead of allocating a new one.

albrow replied:
I updated the code to use a single `bytes.Buffer` for all incoming data. I also made sure that the order of messages is retained.

I'm unsure of how and when to reset the buffer. Any specific suggestions?

Stebalien replied:
I meant:

```go
if c.currData.Len() == 0 {
  // read some data...
  c.currData.Reset(data)
}
```

(but that's no longer an issue).

Stebalien commented on conn_js.go:
```diff
@@ -0,0 +1,196 @@
+		close(openSignal)
+		return nil
+	})
+	defer func() {
```
nit: this can be `defer c.Call("removeEventListener", "open", handler)`.

albrow replied:
👍 fixed.

albrow commented:
@Stebalien I addressed all your feedback. I still need to add more robust error handling. Feel free to take another look now or I'll ping you again after error handling has been improved.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
```
nit: we should probably cache the remote addr/multiaddr.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	return NewAddr("0.0.0.0:0")
```
Let's cache this.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+			// should check binaryType and then decode accordingly.
+			blob := args[0].Get("data")
+			data := readBlob(blob)
+			c.incomingData <- data
```
This will still deliver messages out of order as goroutines can be run in any order.

Stebalien replied:
The idea behind the read loop was that it would prevent writing to this channel from blocking. That plus a significant buffer would get rid of the need for a goroutine.

However, this could also be accomplished by a lock with a small critical section.

Stebalien replied:
I'd go with the lock. It'll be a lot simpler.

albrow replied:
Sorry @Stebalien I meant for this part to *not* be in a goroutine which would retain the order of messages. I just forgot to remove the goroutine.

Let me try a version where we write to a mutex-protected buffer directly inside of `messageHandler`. I agree it is probably the cleanest solution.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+			return err
+		}
+		c.currDataMut.Unlock()
+		c.dataSignal <- struct{}{}
```
This will block until the data is read. It should be a channel-based flag (size-one buffer, write to it with a select).

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+// FileReader API under the hood.
+func readBlob(blob js.Value) []byte {
+	reader := js.Global().Get("FileReader").New()
+	dataChan := make(chan []byte)
```
This should have a buffer of 1 so loadEndFunc doesn't block.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+		dataLen := view.Get("length").Int()
+		data := make([]byte, dataLen)
+		for i := 0; i < dataLen; i++ {
+			data[i] = byte(view.Index(i).Int())
```
This is going to be painfully slow. Is there no other way to get data out of javascript? Even if it involves allocations, getting the browser to give us something we can just read all at once would be faster.

albrow replied:
I just pushed a commit that switches to using the `readAsBinaryString` method of `FileReader`. Need to do a bit more testing to make sure that this works for any arbitrary data and we don't run into encoding issues. I'm somewhat worries because I have run into issues with the `binary` encoding in JavaScript in the past.

For some additional background, the `onmessage` handler in the JavaScript WebSockets API gives us data of type `Blob`. The [Blob API](https://developer.mozilla.org/en-US/docs/Web/API/Blob) in JavaScript is fairly obtuse. AFAICT the only way to read from a `Blob` is to use a [`FileReader`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader). There are a few different ways to use the `FileReader` API. I originally tried using the `readAsText` method, but was running into problems with how JavaScript handles string encoding. That's why I switched to the `readAsArrayBuffer` method which turns the data into an `ArrayBuffer` which can be converted to `Uint8Array`, which is the JavaScript equivalent of `[]byte`. There is a third method that might be faster for us called `readAsBinaryString`. I think it's faster because we can read the data as a `string` all at once and then convert that to `[]byte` in Go without making anymore `syscall/js` calls.

albrow replied:
Unfortunately `readAsBinaryString` doesn't work. You can't convert directly from a binary string in JavaScript to `[]byte` in Go (I suspect JavaScript is doing something wonky with their "binary" encoding).

When testing out the `readAsBinaryString` method in [0x Mesh](https://github.com/0xProject/0x-mesh), I was seeing the following error:

```
failed to negotiate security protocol: unexpected EOF
```

Here's the first couple frames as reference:

Using `readAsArrayBuffer`:

```
00000000  13 2f 6d 75 6c 74 69 73  74 72 65 61 6d 2f 31 2e  |./multistream/1.|
00000010  30 2e 30 0a                                       |0.0.|

00000000  0d 2f 73 65 63 69 6f 2f  31 2e 30 2e 30 0a        |./secio/1.0.0.|

00000000  00 00 00 6c 0a 10 4e 2b  ae 04 e3 d0 11 a6 d9 6a  |...l..N+.......j|
00000010  62 a6 7c 79 6f 0a 12 25  08 02 12 21 03 3f d2 e5  |b.|yo..%...!.?..|
00000020  b5 9a 7c 00 8c 52 e7 23  25 b1 17 36 47 b3 90 89  |..|..R.#%..6G...|
00000030  6b 99 66 0e 56 b1 5c bf  84 9f 52 ee 56 1a 11 50  |k.f.V.\...R.V..P|
00000040  2d 32 35 36 2c 50 2d 33  38 34 2c 50 2d 35 32 31  |-256,P-384,P-521|
00000050  22 0f 41 45 53 2d 32 35  36 2c 41 45 53 2d 31 32  |".AES-256,AES-12|
00000060  38 2a 0d 53 48 41 32 35  36 2c 53 48 41 35 31 32  |8*.SHA256,SHA512|
```

Using `readAsBinaryString`:

```
00000000  13 2f 6d 75 6c 74 69 73  74 72 65 61 6d 2f 31 2e  |./multistream/1.|
00000010  30 2e 30 0a                                       |0.0.|

00000000  0d 2f 73 65 63 69 6f 2f  31 2e 30 2e 30 0a        |./secio/1.0.0.|

00000000  00 00 00 6c 0a 10 c3 b5  c2 b7 c3 8b c2 8d 12 c3  |...l............|
00000010  ac c2 ac c2 b7 c2 92 c3  a1 c2 9b 3a 78 36 59 6a  |...........:x6Yj|
00000020  12 25 08 02 12 21 03 3f  c3 92 c3 a5 c2 b5 c2 9a  |.%...!.?........|
00000030  7c 00 c2 8c 52 c3 a7 23  25 c2 b1 17 36 47 c2 b3  ||...R..#%...6G..|
00000040  c2 90 c2 89 6b c2 99 66  0e 56 c2 b1 5c c2 bf c2  |....k..f.V..\...|
00000050  84 c2 9f 52 c3 ae 56 1a  11 50 2d 32 35 36 2c 50  |...R..V..P-256,P|
00000060  2d 33 38 34 2c 50 2d 35  32 31 22 0f 41 45 53 2d  |-384,P-521".AES-|
00000070  32 35 36 2c 41 45 53 2d  31 32 38 2a 0d 53 48 41  |256,AES-128*.SHA|
00000080  32 35 36 2c 53 48 41 35  31 32                    |256,SHA512|
```

@Stebalien does anything in that data look weird to you? If I'm right and JavaScript is doing something weird with their binary encoding maybe there is an efficient way to correct the encoding and then convert to `[]byte`?

Otherwise, it looks like Go 1.13 is going to come with a new [`js.CopyBytesToGo` function](https://github.com/golang/go/commit/c468ad04177c422534ad1ed4547295935f84743d) that we could use with the `readAsArrayBuffer` method. So we might just have to go with slower manual copying for now and switch to that when Go 1.13 comes out.

Stebalien replied:
> Otherwise, it looks like Go 1.13 is going to come with a new js.CopyBytesToGo function that we could use with the readAsArrayBuffer method. So we might just have to go with slower manual copying for now and switch to that when Go 1.13 comes out.

Sounds good. Yeah, it looks like an encoding issue.

albrow replied:
@Stebalien I just pushed up a new commit that forces the JavaScript WebSockets API to use `ArrayBuffer` instead of `Blob` for the type of incoming messages. This means we don't have to use the `FileReader` API and we can avoid a lot of headaches with trying to turn asynchronous JavaScript code into synchronous Go code. However, we do still have to manually allocate a new `[]byte` and use a for loop to copy the data for now.

albrow commented:
@Stebalien I took a stab at adding proper error handling. It was definitely tricky to work with the event-driven JavaScript API. It also has some other weird quirks like the fact that the `error` event doesn't return any useful information about what the error actually was (you have to use the `close` event for that).

The basic idea is:

1. If the `error` or `close` event triggers while we are trying to open the connection, return an error in `Dial`.
2. If the `error` or `close` event triggers after the connection is already open, first allow `Read` to read any remaining data and then return the error on the next call to `Read`.

albrow commented:
I removed the WIP prefix from the title. @Stebalien I believe I've addressed all your feedback and this PR is ready for final review. Ideally we can also get the tests working in TravisCI.

albrow commented:
@Stebalien just a friendly reminder that I believe I've addressed all your feedback and am waiting for you to take another look. Let me know if I missed anything or if you have any new thoughts.

Stebalien commented on browser_integration_browser_test.go:
```diff
@@ -0,0 +1,56 @@
+
+	// TODO(albrow): This hack is necessary in order to give the reader time to
+	// finish. We should find some way to remove it.
+	time.Sleep(1 * time.Second)
```
Can you expand on this?

albrow replied:
@Stebalien I expanded on this comment. Let me know if this makes sense.

Stebalien replied:
Can we:

1. Close the stream from the browser after reading "pong".
2. Call `stream.Read([]byte{0})` here to wait for the stream to close?

Stebalien commented on websocket_browser.go:
```diff
@@ -0,0 +1,37 @@
+	rawConn := js.Global().Get("WebSocket").New(wsurl)
+	conn := NewConn(rawConn)
+	if err := conn.waitForOpen(); err != nil {
+		return nil, err
```
We probably need to call close here.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if err := c.checkOpen(); err != nil {
```
As far as I can tell, we should be able to just remove this entirely.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+		if err != nil && err != io.EOF {
+			// Return any unexpected errors immediately.
+			return n, err
+		} else if n == 0 || err == io.EOF {
```
`n > 0 && err == io.EOF` is possible. In that case, we'd discard data.

This check should probably be:

```
if err != nil && err != io.EOF {
  return n, err // we read an error (is this even possible?)
}

if n != 0 {
  return n, nil // we read data, return it.
}

// wait for more data.
```

albrow replied:
Ah good point. Documentation for `bytes.Buffer.Read` makes it clear that the only possible error is `io.EOF`. I refactored the code based on your suggestion.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+	c.currDataMut.RUnlock()
+	if n == 0 {
+		if c.firstErr != nil {
+			return 0, c.firstErr
```
What's `firstErr`? If there's a connection _error_ (not just a polite close), we should abort immediately without bothering to read the rest of the buffer.

albrow replied:
@Stebalien Apparently, due to security concerns [most browsers always use `code 1006: connection unexpectedly closed`](https://stackoverflow.com/questions/19304157/getting-the-reason-why-websockets-closed-with-close-code-1006) whenever a WebSocket connection is closed. I don't think there is any way for us to differentiate between a polite close and an unexpected error in the underlying WebSocket connection.

Given that constraint, what do you want to do here?

Stebalien replied:
I'm not seeing anything to suggest that. From the looks of it, 1006 means "something went wrong". We just need to distinguish between that case and "the connection was closed correctly".

Stebalien replied:
Ah. I see https://stackoverflow.com/a/53340067. That's a bug in chrome and there's nothing we can do about it. Where are you seeing the talk about this being a security feature?

albrow replied:
From the link I shared above:

> However, Chrome will rarely report any close code 1006 reasons to the javascript side. This is likely due to client security rules in the WebSocket spec to prevent abusing websocket. (such as using it to scan for open ports on a destination server, or for generating lots of connections for a denial-of-service attack).

Empirically, `1006` is the only error code I've observed from Chrome.

albrow replied:
Hm.. After reading this again, the quote doesn't exactly back up what I said. It just says that *if* there is a code `1006` error there is no way to get additional information. Let me do some more research. Maybe the reason I've been seeing `1006` has to do with how the connection is closed.

albrow replied:
After doing some more testing, I have observed Chrome reporting status code `1005` (also with an empty `reason` field). However, this doesn't really help us much. I can't find any way to produce code `1000` which is what we would expect for a polite close. This might be because of the bug in Chrome that you mentioned or because `gorilla/websocket` isn't sending the appropriate close frame?

There is that comment you shared from StackOverflow that claims Chrome never reports code `1000` but I did find [these tests in Chromium](https://github.com/crosswalk-project/chromium-crosswalk/blob/dd39e4fbbc1e4417cef2f908cc0a42af1e2120aa/third_party/WebKit/LayoutTests/http/tests/websocket/close-code-and-reason-expected.txt) which suggest there are at least some cases where it does report code `1000`.

How should we proceed?

Stebalien replied:
Thanks for looking into this. Turns out, we _aren't_ sending a nice close message: https://github.com/libp2p/go-ws-transport/pull/54

Stebalien replied:
> It sounds like maybe you aren't satisfied with this approach?

I'm fine with an async approach. The network itself is async so there's nothing we can do about that.

My concern is:

1. When we receive an _error_, we should start failing reads/writes ASAP. This doesn't need to be synchronous, I just don't want to bother finishing reading.
2. When we receive an EOF (nice close), we should continue reading to the end.

Note: We're mostly getting into pedantic correctness edge-cases here.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+		return 0, err
+	}
+	uint8Array := js.Global().Get("Uint8Array").New(len(b))
+	for i := 0; i < len(b); i++ {
```
nit:

```go
for i, bi := range b {
  uint8Array.SetIndex(i, bi)
}
```

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	if err := c.checkOpen(); err != nil {
```
Will something go wrong if we drop this check and just try writing? This will race with any `Close` calls anyways.

albrow replied:
@Stebalien according to [this GitHub issue](https://github.com/websockets/ws/issues/1515) browsers do not throw an exception if you try to call `send` on a closed connection. You can verify for yourself by visiting https://www.websocket.org/echo.html and running the following in the JavaScript console:

```
ws = new WebSocket("wss://echo.websocket.org");
ws.onopen = () => {
  ws.close();
  ws.send("test");
  console.log("sent");
};
```

Chrome prints a message to the console but it doesn't throw an exception. Firefox doesn't give any indication that anything is wrong.

I think we have to leave this check in place just so that callers can know if they are trying to send to a closed connection. It _should_ be okay if it's racy because the connection will be detected as closed pretty quickly in any case. What do you think?

Stebalien replied:
Then this _should_ be fine.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	c.signalClose()
+	c.Call("close")
```
I _think_ we should be closing before signaling, but I'm not sure.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+func (c *Conn) releaseHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
```
Should we be setting these to nil?

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,292 @@
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	c.signalClose()
```
We should probably wrap this entire thing in the once. Close should be idempotent.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,311 @@
+			c.signalClose()
+			c.mut.Lock()
+			// Store the error in c.firstErr. It will be returned by Read later on.
+			c.firstErr = errorEventToError(args[0])
```
I see we're taking the lock when setting this but not when reading it. We should either:

1. Take the lock when reading it.
2. Guard it with the channel:
  1. Set it within `signalClose(err)`.
  2. Set it _before_ closing.
  3. Only read it after observing that the close-signaling channel is actually closed.

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,311 @@
+
+// checkOpen returns an error if the connection is not open. Otherwise, it
+// returns nil.
+func (c *Conn) checkOpen() error {
```
Can we just wait for the "close" event to happen then check the `closeSignal` channel after that? That should save us a round-trip through the webasm runtime.

```go
select {
  case <-c.closeSignal:
    return true
  default:
    return false
}
```

Stebalien commented on conn_browser.go:
```diff
@@ -0,0 +1,311 @@
+}
+
+// arrayBufferToBytes converts a JavaScript ArrayBuffer to a slice of bytes.
+func arrayBufferToBytes(buffer js.Value) []byte {
```
So... go 1.13 is out. Want to go ahead and use `CopyBytesToGo` and `CopyBytesToJs`?

albrow replied:
My personal preference would be to release a version that works with Go 1.12 and _then_ include this optimization. There are some speed bumps with upgrading to Go 1.13 so some projects (0x Mesh included) might not have updated yet.

Stebalien replied:
SGTM. Note: libp2p is likely to update to 1.13 before too long as we need it for ed25519+TLS.

albrow commented:
@Stebalien Thanks for your latest feedback and for investigating that graceful close issue. I'm sorry for the delay on my end. At 0x we've been gearing up for our next big protocol upgrade (version 3) so I've had to temporarily de-prioritize WebAssembly/browser integration. I'll get back to this PR as soon as I can.

Stebalien commented:
Np. And you never need to apologize for a delay. If it comes to it, I can address my own feedback and merge.

Thanks for the heads up and the awesome work!

Stebalien commented:
@albrow I rebased this and fixed the last few nits (and got the tests to pass): https://github.com/libp2p/go-ws-transport/pull/55

(we were pretty close and I didn't want it to bitrot)

Stebalien commented:
@albrow I've merged #55. Please let me know if I broke something.
